src/System Application/App/Language/src/LanguageImpl.Codeunit.al

```diff
@@ -36,8 +36,14 @@ codeunit 54 "Language Impl."
 
     procedure GetLanguageIdOrDefault(LanguageCode: Code[10]): Integer;
     var
+        Language: Codeunit Language;
         LanguageId: Integer;
+        Handled: Boolean;
     begin
+        Language.OnGetLanguageIdOrDefault(LanguageCode, LanguageId, Handled);
```

Comment:
@grobyns @Groenbech96 

Our real problem is the fact that we do not know what is set in a report. You always talk about the "customer record", but the data is not only in the customer table, but also in contacts, setup tables, bank accounts and in multiple purchase and sales headers.

To make matters worse, anyone can now use constants in a report, data from third-party tables and much more, thereby overwriting the region with any data from everywhere. The caller has no control, because the api to overwrite these values is not available in Report.SaveAs().

In fact, I still expect an exported XML (data interchange format!) to contain decimals and date/time values in US English, which is to be corrected in the platform. However, manipulating the outcome from the application is also an important step forward.

These events are too general and, as written above, they lack context. The context may also be lost with Bind/UnbindSubscription().

Would you be open for another approach? Both Language and Language Impl. are already single instance Codeunits. What about the following methods?

- SetOverrideFormatRegion('en-US') -> Override all subsequent calls to GetFormatRegionOrDefault() until set back to empty or another value.
- SetOverrideNextFormatRegion('en-US') -> Override the next call to GetFormatRegionOrDefault(), automatically resets the override.

Same for Language.

So a developer can easily do this

```AL
Language.SetOverrideFormatRegion('en-US');
Report.SaveAs(...);
Language.SetOverrideFormatRegion(); // Overload without param. or Language.ResetOverrideFormatRegion()

// Handle proper XML output
```

What do you think? I know, setting overrides and not resetting it can cause massive issues, but filtering on FilterGroup 1 can do too...

Reply:
The problem with using this kind of override is that if an error happens, the incorrect value is set until the user closes the session and starts a new one. 
With manual binding, if an error happens, the unbinding should happen when the codeunit instance goes out of scope which is much more controllable. (i.e. don't use manual binding in a single instance cu and you should be fine).

I wonder if there is a way to combine some of these ideas such that these events are only raised when needed and subscribers are only called when needed and everything would automatically reset on errors to avoid unwanted after effects... it's still not as good as fixing the platform but would minimize issues...

Reply:
Thinking a bit more about the SetOverride... you imagine something like this i assume:

 ```suggestion
    procedure GetLanguageIdOrDefault(LanguageCode: Code[10]): Integer;
    var
        LanguageId: Integer;
    begin
        if LanguageIdOverride <> 0 then begin
            LanguageId := LanguageIdOverride;
            LanguageIdOverride := 0;
            exit(LanguageId);
        end;
        
        LanguageId := GetLanguageId(LanguageCode);

        if LanguageId = 0 then
            LanguageId := GlobalLanguage();

        exit(LanguageId);
    end;
```

same for FormatRegion. it would act exactly once and then reset. So you would never need to manually reset afterwards. If an error occurs after you set the override but before the call to GetLanguage then whatever is the next call would get it... This does seem to minimize any potential fallout so that's ok... not sure we can do much better as an app fix actually...

Reply:
Yes, exactly (for the idea with SetOverrideNextFormatRegion()). Let me check another option, not sure if this works. Gimme some minutes.

Reply:
While I could go with the shot and forget approach, there might be another option. But I have no clue if this is something you would do:

```AL
if not IsOnGetFormatRegionOrDefaultStaticallyBound() then begin
    OnGetFormatRegionOrDefaultManualBinding(FormatRegion);
end;
```

```AL
local procedure IsOnGetFormatRegionOrDefaultStaticallyBound(): Boolean
var
    EventSubscription: Record "Event Subscription";
begin
    EventSubscription.SetRange("Publisher Object Type", EventSubscription."Publisher Object Type"::Codeunit);
    EventSubscription.SetRange("Publisher Object ID", Codeunit::Language);
    EventSubscription.SetRange("Published Function", 'OnGetFormatRegionOrDefault');
    EventSubscription.SetRange("Subscriber Instance", 'Static-Automatic');

    exit(not EventSubscription.IsEmpty());
end;

[IntegrationEvent(false, false)]
internal procedure OnGetFormatRegionOrDefaultManualBinding(var FormatRegion: Text[80])
begin
end;
```

A bit tricky but would allow you to only allow manual even binding. Also, it could be initialized on the first call, changes would only occur if apps are changed.

Your decision, one time override or more complex checks for event bindings.

Reply:
I like the idea of checking for manual binding but not the implementation of it. This really should just be an attribute on the event publisher which states what kind of binding is allowed and the compiler should enforce it... 

I'll pass the idea on to the compiler team but for the app workaround I think we should go with the one-time override for now.

Reply:
Sounds good to me, thank you!

Reply:
I just wanted to have that mentioned here. If the platform in the future supports setting FormatRegion and Language for Report.SaveAs() programmatically like Report.Run() does (in the page, interactively), then no changes regarding the xml are needed.

Reply:
Adding these two parameters in a parameter bag is included in current ideas. The report should be able to run without depending on triggers, request page xml (which is limited to page controls/filters) etc. to ensure a well-defined output without user or app interaction.

Reply:
Please review changes to PR. API now has 2 override methods.

Reply:
Looks good to me, thank you.